src/module.c

```diff
@@ -10388,6 +10393,154 @@ int RM_GetDbIdFromDefragCtx(RedisModuleDefragCtx *ctx) {
+    UNUSED(ae);
+    EventLoopData *data = user_data;
+    data->wProc(fd, data->user_data, eventLoopFromAeMask(ae_mask));
+}
+
+/* Add a pipe / socket event to the event loop
+ *
+ * On success REDISMODULE_OK is returned, otherwise
+ * REDISMODULE_ERR is returned and errno is set to the following values:
+ *
+ * * ERANGE: 'fd' is negative or higher than 'maxclients' Redis config.
+ * * EINVAL: 'callback' is NULL or 'mask' value is invalid.
+ *   errno might take other values in case of an internal error.
+ *
+ *  Example:
+ *  void onReadable(int fd, void *user_data, int mask) {
+ *      char buf[32];
+ *      int bytes = read(fd,buf,sizeof(buf));
+ *      printf("Read %d bytes \n", bytes);
+ *  }
+ *  RM_EventLoopAdd(fd, REDISMODULE_EVENTLOOP_READABLE, onReadable, NULL);
+ */
+int RM_EventLoopAdd(int fd, int mask, RedisModuleEventLoopCallback callback, void *user_data) {
+    if (fd < 0 || fd >= aeGetSetSize(server.el)) {
+        errno = ERANGE;
+        return REDISMODULE_ERR;
+    }
+
+    if (!callback || mask & ~(REDISMODULE_EVENTLOOP_READABLE |
+                              REDISMODULE_EVENTLOOP_WRITABLE)) {
+        errno = EINVAL;
+        return REDISMODULE_ERR;
+    }
+
+    /* We are going to register stub callbacks to 'ae' for two reasons :
+     * - "ae" callback signature is different from RedisModuleEventLoopCallback,
+     *   that will be handled it in our stub callbacks.
+     * - We need to remap 'mask' value to provide binary compatibility.
+     *
+     *  For the stub callbacks, saving user 'callback' and 'user_data' in an
+     *  EventLoopData object and passing it to ae, later we'll extract
+     *  'callback' and 'user_data' from that.
+     */
+    EventLoopData *data = aeGetFileClientData(server.el, fd);
+    if (!data)
+        data = zcalloc(sizeof(*data));
+
+    aeFileProc *aeProc;
+    if (mask & REDISMODULE_EVENTLOOP_READABLE)
+        aeProc = eventLoopCbReadable;
+    else
+        aeProc = eventLoopCbWritable;
+
+    int aeMask = eventLoopToAeMask(mask);
+
+    if (aeCreateFileEvent(server.el, fd, aeMask, aeProc, data) != AE_OK) {
+        if (aeGetFileEvents(server.el, fd) == AE_NONE)
+            zfree(data);
+        return REDISMODULE_ERR;
+    }
+
+    data->user_data = user_data;
+    if (mask & REDISMODULE_EVENTLOOP_READABLE)
+        data->rProc = callback;
+    if (mask & REDISMODULE_EVENTLOOP_WRITABLE)
+        data->wProc = callback;
+
+    errno = 0;
+    return REDISMODULE_OK;
+}
+
+/* Delete a pipe / socket event from the event loop.
+ *
+ * On success REDISMODULE_OK is returned, otherwise
+ * REDISMODULE_ERR is returned and errno is set to the following values:
+ *
+ * * ERANGE: 'fd' is negative or higher than 'maxclients' Redis config.
+ * * EINVAL: 'mask' value is invalid.
+ *            Valid values for 'mask':
+ *               REDISMODULE_EVENTLOOP_READABLE
+ *               REDISMODULE_EVENTLOOP_WRITABLE
+ *               REDISMODULE_EVENTLOOP_READABLE | REDISMODULE_EVENTLOOP_WRITABLE
```

Comment:
Sub-list indented 4 spaces more that the previous level. Empty line before it.

```suggestion
 * * EINVAL: 'mask' value is invalid. Valid values for 'mask':
 *
 *     * `REDISMODULE_EVENTLOOP_READABLE`
 *     * `REDISMODULE_EVENTLOOP_WRITABLE`
 *     * `REDISMODULE_EVENTLOOP_READABLE | REDISMODULE_EVENTLOOP_WRITABLE`
```